src/k2/tso/service/Service.cpp

```diff
@@ -0,0 +1,220 @@
+
+    The above copyright notice and this permission notice shall be included in all copies
+    or
+    substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS",
+    WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+    DAMAGES OR OTHER
+    LIABILITY,
+    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+*/
+#include <k2/appbase/Appbase.h>
+#include <k2/dto/MessageVerbs.h>
+#include <k2/dto/Timestamp.h>
+
+#include "Exceptions.h"
+#include "Service.h"
+
+namespace k2::tso {
+
+TSOService::TSOService() {
+    K2LOG_I(log::tsoserver, "ctor");
+}
+
+seastar::future<> TSOService::start() {
+    K2LOG_I(log::tsoserver, "TSOService start");
+    _metricGroups.clear();
+
+    std::vector<sm::label_instance> labels;
+
+    _metricGroups.add_group("TSO", {
+        sm::make_histogram("timetamp_error", [this]{ return _timestampErrors.getHistogram();},
+                sm::description("Errors in returned timestamp in nanoseconds"), labels)
+    });
+
+    seastar::future<> startFut = seastar::make_ready_future<>();
+    if (seastar::this_shard_id() == 0) {
+        K2LOG_I(log::tsoserver, "TSOService initializing GPS clock");
+        GPSClock::initialize();
+        auto now = GPSClock::now();
+        K2LOG_I(log::tsoserver, "TSOService clock has been initialized with GPS steady={}, real={}, error={}",
+                     nsec(now.steady).count(), nsec(now.real).count(), nsec(now.error).count());
+        if (now.error > _errorBound()) {
+            K2LOG_E(log::tsoserver, "TSOService cannot start since the GPS error is larger than our error bound {}", _errorBound());
+            return seastar::make_exception_future<>(std::runtime_error("gps error is bigger than max error bound"));
+        }
+        // start the poller thread
+        K2LOG_I(log::tsoserver, "Starting GPS clock poller on CPU: {}", _clockPollerCPU());
+        _keepRunningPoller.test_and_set(); // set the running flag to true
+        _clockPoller = std::thread([this, pinCPU=_clockPollerCPU()] {
+            // pin the calling thread to the given CPU
+            if (pinCPU >= 0) {
+                cpu_set_t cpuset;
+                CPU_ZERO(&cpuset);
+                CPU_SET(pinCPU, &cpuset);
+                if (0 != pthread_setaffinity_np(pthread_self(), sizeof(cpu_set_t), &cpuset)) {
+                    throw std::runtime_error("Unable to set affinity");
+                }
+            }
+            while (_keepRunningPoller.test_and_set()) {

```

Comment:
why do we need to call test_and_set() to set the flag to true here? shouldn't we just check if the flag is true and stop polling when the flag is false?